hoomd/hpmc/PairPotentialUnion.h

```diff
@@ -0,0 +1,284 @@
+        // r_cuts over type pairs that interact between two union particles of type_i and type_j
+
+        LongReal r_cut = 0;
+
+        for (auto& constituent_type_i : m_type[type_i])
+            {
+            for (auto& constituent_type_j : m_type[type_j])
+                {
+                r_cut
+                    = std::max(r_cut,
+                               m_constituent_potential->computeRCutNonAdditive(constituent_type_i,
+                                                                               constituent_type_j));
+                }
+            }
+
+        return r_cut;
+        }
+
+    /// Returns the additive part of the cutoff distance for a given type.
+    virtual LongReal computeRCutAdditive(unsigned int type) const
+        {
+        // The additive cutoff is twice the radius of the constituent particle furthest from the
+        // origin.
+        assert(type <= m_extent_type.size());
+
+        if (m_constituent_potential->computeRCutAdditive(type) > 0)
+            {
+            throw std::domain_error("Unsupported constituent potential.");
+            }
+
+        return m_extent_type[type];
+        }
+
+    protected:
+    /// The pair potential to apply between constituents.
+    std::shared_ptr<PairPotential> m_constituent_potential;
+
+    /// The tree acceleration structure per particle type.
+    std::vector<hpmc::detail::GPUTree> m_tree;
+
+    /// The per-type geometric extent.
+    std::vector<Scalar> m_extent_type;
+
+    /// The positions of the constituent particles.
+    std::vector<std::vector<vec3<LongReal>>> m_position;
+
+    // The orientations of the constituent particles.
+    std::vector<std::vector<quat<LongReal>>> m_orientation;
+
+    // The charges of the constituent particles.
+    std::vector<std::vector<LongReal>> m_charge;
+
+    // The type identifiers of the constituent particles.
+    std::vector<std::vector<unsigned int>> m_type;
+
+    // The number of particles in a leaf of the internal tree data structure.
+    unsigned int m_leaf_capacity;
+
+    /// Update extent.
+    void updateExtent(unsigned int type_id);
+
+    /// Builds OBB tree based on geometric properties of the constituent particles.
+    void buildOBBTree(unsigned int type_id);
+
+    /// Compute the energy of two overlapping leaf nodes.
+    LongReal compute_leaf_leaf_energy(vec3<LongReal> dr,
+                                      unsigned int type_a,
+                                      unsigned int type_b,
+                                      const quat<LongReal>& orientation_a,
+                                      const quat<LongReal>& orientation_b,
+                                      unsigned int cur_node_a,
+                                      unsigned int cur_node_b) const;
+
+    /*** Evaluate the energy of the pair interaction (all N * M pairs).
+     */
+    __attribute__((always_inline)) LongReal energyAll(const LongReal r_squared,
```

Comment:
I would prefer either all long functions be defined in the source file or all in the header.

Reply:
These are the "short" functions that need to be inlined for performance.

Reply:
I see, there were some other "long" functions in the header. I move those all to the .cc file.